### DISCUSSION TITLE
[Bug]: Unable to download assistant files

### DISCUSSION THREAD

**Discussion by VS-KerryFloyd** [danny-avila/LibreChat#2358] [Troubleshooting]
### What happened?

Running on the very latest dev release, I'm not able to download assistant generated files. An error appears at the top of the screen and says "Error downloading file. The file may have been deleted."

Opening the download link in a new tab returns 401 status code (Unauthorized).

### Steps to Reproduce

1) Ask assistant to "Generate a json file containing test data for a contacts database"
2) Attempt to download the generated file

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by danny-avila** [comment]
I’ll look into it, seems to be some odd behavior happening since it was first implemented

**Reply by VS-KerryFloyd** [reply]
Thank you!

**Reply by danny-avila** [reply]
@kezzzsVS it works locally, so may be an issue only on remote servers, is this the case for you?

**Reply by VS-KerryFloyd** [reply]
I have this running as a container in Azure. Did you want me to check any particular file or folder permissions within the container?

**Reply by danny-avila** [reply]
are you serving it with NGINX? Remote servers can have strict default policies on serving file downloads, it has nothing to do with file permissions since the file data is streamed from OpenAI

**Reply by danny-avila** [reply]
I'm also trying some things on the code side and see if this can be alleviated.

**Reply by VS-KerryFloyd** [reply]
Requests are going through the Azure's Load Balancer they place in front of containers, however that's non configurable. I've tested this using several different extensions including txt, htm and html. Each of them have the same result unfortunately.
Curiously, having dev tools opens reports a 404 rather than 401 when left clicking the link from the text. The reported request url appears to be both partially URL encoded and different (/api/files/.... vs /api/files/download/....)

For example, right clicking the link and opening in a new tab has the URL:
https://dev.MYURL.com/api/files/15fd662a44f153155c59d1b1/assistant-EC1Ng0wAvmyHDQBq6C885FPu/contacts.json
which results in a 401.

Left clicking the link reports the following attempt on the Network tab of the dev tools which results in a 404:
https://dev.MYURL.com/api/files/download/15fd662a44f153155c59d1b1/files%2F65fd662a44f553155c59d1a1%2Fassistant-EC1Ng0wAvmyHDQBq6C885FPu%2Fcontacts.json

**Reply by danny-avila** [reply]
I'm having trouble making this work on cloudflare + NGINX as well. I hit 404 as something blocks it altogether, the API never receives the requests.

401 makes sense, because navigating to the link direct won't authenticate the request.

I am getting to the bottom of this because it seems that the request URI is being stripped and not mapped by the default nginx proxy_pass

I can potentially solve this for nginx however:

> Azure's Load Balancer they place in front of containers, however that's non configurable

since you have no configuration option, maybe changing the implementation from a GET query to POST mutation is the one size fits all, though this isn't common or best practice

**Reply by danny-avila** [reply]
For the record, switching nginx config from

```bash
  location /api {
    proxy_pass http://api:3080/api;
  }
```

to

```bash
  location /api/ {
    proxy_pass http://api:3080$request_uri;
  }
```

does fix the issue.

Unsure how this would work with Azure Containers

**Reply by sathyarajv** [reply]
For Azure Containers, I made these two changes to resolve this issue. 

In [/client/src/components/Chat/Messages/Content/Markdown.tsx](https://github.com/danny-avila/LibreChat/blob/6f0eb3536526e0bb4a9455dabd983b27fd111fd9/client/src/components/Chat/Messages/Content/Markdown.tsx)

```
const { filepath, filename } = useMemo(() => {
    const pattern = new RegExp(`(?:files|outputs)/${user?.id}/([^\\s]+)`);
    const match = href.match(pattern);
    if (match && match[0]) {
      const path = match[0];
      const name = path.split('/').pop();
      return { filepath: path.split('/')[2], filename: name }; //passing only fileid to handle route issue in api
    }
    return { filepath: '', filename: '' };
  }, [user?.id, href]);
```

In [/api/server/routes/files/files.js](https://github.com/danny-avila/LibreChat/blob/6f0eb3536526e0bb4a9455dabd983b27fd111fd9/api/server/routes/files/files.js)
![image](https://github.com/danny-avila/LibreChat/assets/9957258/77605695-f779-4915-aa46-595d6946af1a)


Passing only the file Id instead of whole path resolves this issue.

**Reply by danny-avila** [reply]
@sathyarajv makes sense, I can adapt this

**Comment by danny-avila** [comment]
This is now resolved here: https://github.com/danny-avila/LibreChat/pull/2364

**Comment by VS-KerryFloyd** [comment]
Thanks for the super quick response on this! I'm afraid it doesn't seem to be working for me, although the error has changed.

Upon clicking a download, the dev tools will now report a 500 Internal Server Error for the following URL:
https://dev.MYURL.com/api/files/download/98fd662a44f113155c59d1a1/assistant-AQr7e7LKA91Gu4s8hYJ7XLaI
In the Javascript console, an "Error downloading file: No data found" will appear.
I have tried generating new files, but same result.

**Reply by danny-avila** [reply]
possibly specific to Azure, will look into it.

**Reply by VS-KerryFloyd** [reply]
Hi Danny - thank you. Did you need me to re-open this or anything?

**Reply by danny-avila** [reply]
@kezzzsVS if you'd like to with the exact errors you're facing now, feel free. I plan on taking care of this by end of day Friday.

**Reply by VS-KerryFloyd** [reply]
@danny-avila - I've just tested it today and it's working perfectly. I'm not sure what the issue was before, or if you made some additional changes, but no need for any further work from what I've tested. Thank you very much!

**Reply by danny-avila** [reply]
@kezzzsVS thanks for letting me know, have not made any relevant changes that I know of! but I'm glad it's working
